src/Tests/Microsoft.Diagnostics.Monitoring.Tool.UnitTests/ExecuteActionTests.cs

```diff
@@ -0,0 +1,169 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.Diagnostics.Tools.Monitor.CollectionRules.Options.Actions;
+using System.Threading.Tasks;
+using Xunit;
+using Microsoft.Diagnostics.Tools.Monitor.CollectionRules.Actions;
+using System.Reflection;
+using Microsoft.Diagnostics.Monitoring.TestCommon;
+using System.Threading;
+using System;
+using System.IO;
+using System.Diagnostics;
+using Microsoft.Diagnostics.Tools.Monitor;
+
+namespace Microsoft.Diagnostics.Monitoring.Tool.UnitTests
+{
+    public sealed class ExecuteActionTests
+    {
+        private const int TokenTimeoutMs = 10000;
+        private const int DelayMs = 1000;
+
+        [Fact]
+        public async Task ExecuteAction_ZeroExitCode()
+        {
+            ExecuteAction action = new();
+
+            ExecuteOptions options = new();
+
+            options.Path = DotNetHost.HostExePath;
+            options.Arguments = GenerateArgumentsString(new string[] { "ZeroExitCode" });
+
+            using CancellationTokenSource cancellationTokenSource = new CancellationTokenSource(TokenTimeoutMs);
+
+            CollectionRuleActionResult result = await action.ExecuteAsync(options, null, cancellationTokenSource.Token);
+
+            ValidateActionResult(result, "0");
+        }
+
+        [Fact]
+        public async Task ExecuteAction_NonzeroExitCode()
+        {
+            ExecuteAction action = new();
+
+            ExecuteOptions options = new();
+
+            options.Path = DotNetHost.HostExePath;
+            options.Arguments = GenerateArgumentsString(new string[] { "NonzeroExitCode" });
+
+            using CancellationTokenSource cancellationTokenSource = new CancellationTokenSource(TokenTimeoutMs);
+
+            InvalidOperationException invalidOperationException = await Assert.ThrowsAsync<InvalidOperationException>(
+                () => action.ExecuteAsync(options, null, cancellationTokenSource.Token));
+
+            Assert.Contains(string.Format(Strings.ErrorMessage_NonzeroExitCode, "1"), invalidOperationException.Message);
+        }
+
+        [Fact]
+        public async Task ExecuteAction_TokenCancellation()
+        {
+            ExecuteAction action = new();
+
+            ExecuteOptions options = new();
+
+            options.Path = DotNetHost.HostExePath;
+            options.Arguments = GenerateArgumentsString(new string[] { "Sleep", (TokenTimeoutMs + DelayMs).ToString() }); ;
```

Comment:
This might work most of the time. If it's not stable enough, maybe we can change this to write out a file (like the TextFileOutput scenario) and then sleep for some large amount of time. Back in the test, we start the process, wait for the file to appear, and then immediately cancel the token. This would provide more determinism so that if the host happens to be really slow that process startup doesn't exceed the token timeout.